kernel/src/net/socket/ip/raw/mod.rs

```diff
@@ -0,0 +1,427 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use aster_bigtcp::{
+    socket::{SocketEventObserver, SocketEvents},
+    wire::{IpAddress, IpEndpoint},
+};
+use ostd::sync::LocalIrqDisabled;
+use takeable::Takeable;
+
+use self::{bound::BoundRaw, unbound::UnBoundRaw};
+use super::{common::get_ephemeral_endpoint, UNSPECIFIED_LOCAL_ENDPOINT};
+use crate::{
+    events::IoEvents,
+    fs::{
+        file_handle::FileLike,
+        utils::{InodeMode, Metadata, StatusFlags},
+    },
+    match_sock_option_mut,
+    net::{
+        iface::IfaceEx,
+        socket::{
+            options::{Error as SocketError, IpHdrIncl, SocketOption},
+            util::{
+                options::{IpSocketOptionSet, SocketOptionSet},
+                send_recv_flags::SendRecvFlags,
+                socket_addr::SocketAddr,
+                MessageHeader,
+            },
+            Socket,
+        },
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    util::{net::Protocol, MultiRead, MultiWrite},
+};
+
+mod bound;
+mod unbound;
+
+#[derive(Debug, Clone)]
+struct OptionSet {
+    socket: SocketOptionSet,
+    ip: IpSocketOptionSet,
+}
+
+impl OptionSet {
+    fn new() -> Self {
+        let socket = SocketOptionSet::new_raw();
+        let ip = IpSocketOptionSet::new_raw();
+        OptionSet { socket, ip }
+    }
+}
+
+pub struct RawDGramSocket {
+    options: RwLock<OptionSet>,
+    inner: RwLock<Takeable<Inner>, LocalIrqDisabled>,
+    nonblocking: AtomicBool,
+    pollee: Pollee,
+}
+
+enum Inner {
+    Unbound(UnBoundRaw),
+    Bound(BoundRaw),
+}
+
+impl Inner {
+    fn bind(
+        self,
+        addr: &IpAddress,
+        can_reuse: bool,
+    ) -> core::result::Result<BoundRaw, (Error, Self)> {
+        let unbound_raw = match self {
+            Inner::Unbound(unbound_raw) => unbound_raw,
+            Inner::Bound(bound_raw) => {
+                return Err((
+                    Error::with_message(Errno::EINVAL, "the socket is already bound to an address"),
+                    Inner::Bound(bound_raw),
+                ));
+            }
+        };
+
+        let bound_raw = match unbound_raw.bind(addr, can_reuse) {
+            Ok(bound_raw) => bound_raw,
+            Err((err, unbound_raw)) => return Err((err, Inner::Unbound(unbound_raw))),
+        };
+        Ok(bound_raw)
+    }
+
+    fn bind_to_ephemeral_endpoint(
```

Comment:
I don't think raw sockets need an ephemeral port. I believe that raw sockets have nothing to do with the port.

```rust
        // Raw socket is no need to release port
```
As you said, raw sockets will not free the port, so it should not be bound to a port either, otherwise the port can never be freed.